### ISSUE TITLE
Bug: no-unused-vars triggered with typescript this: typing

### ISSUE THREAD

**Issue by alzalabany** [eslint/eslint#15457]
### Environment

Node version:  14.*
Local ESLint version: ^8.5.0
Global ESLint version: none
Operating System: macOS

![image](https://user-images.githubusercontent.com/3620349/147417124-927620d6-cc09-44ac-b53d-bca71829cbe9.png)

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration using "eslint-config-xo"</summary>

```
'no-unused-vars': [
			'error',
			{
				vars: 'all',
				args: 'after-used',
				ignoreRestSiblings: true,
				argsIgnorePattern: /^_/.source, // as a temp fix i now use argsIgnorePattern: /^(_|(this$))/.source,
				caughtErrors: 'all',
				caughtErrorsIgnorePattern: /^_$/.source
			}
		],
```
</details>

```js
export function metrics(this: UserConnection) {
  this.emit("hello", "world");
}
```

### What did you expect to happen?

this was used, so there should be no error.

### What actually happened?

it show unused error

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by B2o5T** [comment]
Please read official documentation about this rule in typescript-eslint https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-unused-vars.md

Original ESLint rule should be disabled in TypeScript environment as it can produce false-positive errors.
